subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c

```diff
@@ -0,0 +1,400 @@
+{
+	struct backend_config_t *conf = priv;
+
+	if (conf) {
+		mbox_send(&conf->mbox_tx, NULL);
+	}
+}
+
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *data;
+	unsigned int vq_id;
+
+	data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	vq_id = (data->role == VIRTIO_DEV_MASTER) ? VIRTQUEUE_ID_PRIMARY : VIRTQUEUE_ID_SECONDARY;
+
+	virtqueue_notification(data->vr.vq[vq_id]);
+}
+
+static void mbox_callback(const struct device *instance, uint32_t channel,
+			  void *user_data, struct mbox_msg *msg_data)
+{
+	struct backend_data_t *data = user_data;
+
+	k_work_submit_to_queue(&mbox_wq, &data->mbox_work);
+}
+
+static int mbox_init(const struct device *instance)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *data = instance->data;
+	int err;
+
+	k_work_queue_start(&mbox_wq, mbox_stack, K_KERNEL_STACK_SIZEOF(mbox_stack),
+			   WQ_PRIORITY, NULL);
+
+	k_work_init(&data->mbox_work, mbox_callback_process);
+
+	err = mbox_register_callback(&conf->mbox_rx, mbox_callback, data);
+	if (err != 0) {
+		return err;
+	}
+
+	return mbox_set_enabled(&conf->mbox_rx, 1);
+}
+
+static int register_ept(const struct device *instance, void **token,
+			const struct ipc_ept_cfg *cfg)
+{
+	struct backend_data_t *data = instance->data;
+	struct ipc_rpmsg_instance *rpmsg_inst;
+	struct ipc_rpmsg_ept *rpmsg_ept;
+	bool ept_cached;
+	int err;
+
+	if (!data->inited) {
+		return -ENODEV;
+	}
+
+	rpmsg_inst = &data->rpmsg_inst;
+
+	k_mutex_lock(&rpmsg_inst->mtx, K_FOREVER);
+	ept_cached = get_ept(rpmsg_inst, &rpmsg_ept, cfg->name);
+
+	if (rpmsg_ept == NULL) {
+		k_mutex_unlock(&rpmsg_inst->mtx);
+		return -ENOENT;
+	}
+
+	rpmsg_ept->cb = &cfg->cb;
+	rpmsg_ept->priv = cfg->priv;
+	rpmsg_ept->bound = false;
+	rpmsg_ept->ep.priv = rpmsg_ept;
+
+	if (ept_cached) {
+		/*
+		 * The endpoint was cached in the NS bind callback. We can finally
+		 * advertise it.
+		 */
+		k_mutex_unlock(&rpmsg_inst->mtx);
+		advertise_ept(rpmsg_inst, rpmsg_ept, cfg->name, rpmsg_ept->dest);
+	} else {
+		/*
+		 * There is no endpoint in the cache because the SECONDARY has
+		 * not registered the endpoint yet. Cache it.
+		 */
+		strncpy(rpmsg_ept->name, cfg->name ? cfg->name : "", sizeof(rpmsg_ept->name));
+		k_mutex_unlock(&rpmsg_inst->mtx);
+	}
+
+	err = ipc_rpmsg_register_ept(rpmsg_inst, data->role, rpmsg_ept);
+	if (err != 0) {
+		return err;
+	}
+
+	(*token) = rpmsg_ept;
+
+	return 0;
+}
+
+static int send(const struct device *instance, void *token,
+		const void *data, size_t len)
+{
+	struct ipc_rpmsg_ept *rpmsg_ept;
+
+	rpmsg_ept = (struct ipc_rpmsg_ept *) token;
+
+	return rpmsg_send(&rpmsg_ept->ep, data, len);
+}
+
+static int open(const struct device *instance)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *data = instance->data;
+	struct ipc_rpmsg_instance *rpmsg_inst;
+	int err = 0;
+
+	if (!atomic_cas(&data->inited, false, true)) {
+		return -EALREADY;
```

Comment:
It looks one of the threads can get `-EALREADY` before the other thread finishes initialization of given instance. Should we take care of such possible race condition?

Reply:
uhm, unlikely but not impossible. Two possible solutions here:
1. Implementing a state machine `READY->BUSY->INITED` for the instance but this means that we should explicitly busy-wait on the `open()` function when getting `BUSY`
2. Protecting the whole `open` code with a mutex / spinlock

Reply:
Implemented a `READY->BUSY->INITED` state machine